### PR TITLE
WIP: Alternative to #7920: Round depth in the fragment shader. 

### PR DESCRIPTION
**Pull request by hrydgard**
Simpler alternative to #7920 but will not work on ES 2.0, and will in many cases be slower. Therefore, we only enable it in the one game where it works much better than #7920, Heroes Phantasia.

OpenGL-only.

### CODE REVIEW

**Comment by hrydgard** [comment]
With ARB_conservative_depth this could even be faster than the vertex shader solution, but it's only available on GLSL 4.20 or with the extension.

https://www.opengl.org/registry/specs/ARB/conservative_depth.txt

**Comment by unknownbrackets** [comment]
Well, we don't know if it'll round up or down, right?  I guess we could say it won't change it, but then it might be wrong.

In God of War, I lose 4% speed as compared to #7920 with this, and that's at 1x.  However, importantly, #7920 is still giving me incorrect depth values still (38022.99 etc.) whereas this gives me accurate ones.

-[Unknown]

**Comment by LunaMoo** [comment]
Big improvement in Heroes Phantasia flickers just like once per second instead every frame and mostly later in the game,  values still differ like for example: 15976.999023 and 15975.999023.
 The difference is whole 1.0, so maybe needs some slight offset before floor like the vertex one? I tried adding 0.125, 0.350 and 0.5 but neither helped.
 I might check more offset values later, unless offset is also wrong solution like bugging Photo Kano in the vertex rounding branch.
  //Moo

**Comment by hrydgard** [comment]
@unknownbrackets Surely floor() will only round down, right?

**Comment by hrydgard** [comment]
I think I'll test tonight how it does with ARB_conservative_depth - if we can run this with no speed loss, I'll just enable this by default where available (and in that case, we disable #7920). I don't think this is well supported on mobile though, so we'll still need #7920 on many platforms.

**Comment by sum2012** [comment]
Test Phantasy Star Portable 2 JP fixed missing word

**Comment by unknownbrackets** [comment]
Oh, I was still thinking about the rounding up at 0.625.  Well, I guess it's fine without.

-[Unknown]

**Comment by hrydgard** [comment]
This seems to be about a 10% speed penalty for me, even with conservative_depth enabled, while the vertex shader method is more like 1%. Of course this will depend on the game, etc...

**Comment by sum2012** [comment]
I take a search.
The VPOS semantic would be the HLSL equivalent of gl_FragCoord.
http://www.gamedev.net/topic/544071-getting-screenviewport-coordinates-of-current-pixel-in-pixel-shader/?view=findpost&p=4507872

**Comment by daniel229** [comment]
This branch does not fix for Taiko no Tatsujin DX,https://github.com/hrydgard/ppsspp/pull/7920 does.

File: GPU/GLES/FragmentShaderGenerator.cpp [hunk 1]
```diff
@@ -1106,6 +1106,13 @@ void GenerateFragmentShader(char *buffer) {
 		WRITE(p, "  %s = vec4(1,0,1,1);\n", fragColor0);
 	}
 #endif
+
+	if (true) {
```
Comment (InusualZ) [inline GPU/GLES/FragmentShaderGenerator.cpp]: why put it in a if statement if always if going to be true?

Comment (FireyFly) [inline GPU/GLES/FragmentShaderGenerator.cpp]: > (the "true" would be replaced with a config option or detection or something..)

Comment (InusualZ) [inline GPU/GLES/FragmentShaderGenerator.cpp]: Oh sorry.

**Pull request merged**